### ISSUE TITLE
Login notifications (code included)

### ISSUE THREAD

**Issue by smkiewel** [ideashower/ShareKit#213]
I needed notifications of when a login had successfully occurred, so I've written that feature. The below code will post notifications to the NotificationCenter with the name @"SHKLogin", with the userInfo dictionary containing the class name of the sharer logged in under the key @"sharerId". 

I believe I've put them in the best places to ensure it gets called only when a successful login event occurs, but if there is a better place, please let me know. 

This requires changes to both SHKSharer and SHKOAuthSharer classes, as well as to the SHKFacebook class, as it uses a different login mechanism.

I have tested this with Delicious, Facebook (not with proxy), Google Reader, Pinboard, Read It Later, Tumblr, and Twitter. I did not test Evernote, as I have not yet gotten that sharer working.

EDIT: DRYed code.

SHKSharer.h:

```
-(void)postAuthorizationNotification;
```

SHKSharer.m:

```
-(void)postAuthorizationNotification {
    NSDictionary *userInfo = [NSDictionary dictionaryWithObject:[self sharerId] forKey:@"sharerId"];
    NSNotification *note = [NSNotification notificationWithName:@"SHKLogin" object:self userInfo:userInfo];
    [[NSNotificationCenter defaultCenter] postNotification:note];
} 

- (void)authorizationFormSave:(SHKFormController *)form {
    // -- Save values 
    NSDictionary *formValues = [form formValues];

    NSString *value;
    NSString *sharerId = [self sharerId];
    NSArray *fields = [[[form sections] objectAtIndex:0] objectForKey:@"rows"];
    for (SHKFormFieldSettings *field in fields) {
        value = [formValues objectForKey:field.key];
        [SHK setAuthValue:value forKey:field.key forSharer:sharerId];
    }

    [self postAuthorizationNotification];    

    // -- Try to share again
    if ( [self validateItem] )  //only share if we have an item -  wont have item if only logging in
        [self share];
}
```

SHKOAuthSharer.m: 

```
- (void)storeAccessToken {
    [SHK setAuthValue:accessToken.key
               forKey:@"accessKey"
            forSharer:[self sharerId]];

    [SHK setAuthValue:accessToken.secret
               forKey:@"accessSecret"
            forSharer:[self sharerId]];

    [SHK setAuthValue:accessToken.sessionHandle
               forKey:@"sessionHandle"
            forSharer:[self sharerId]];

    [self postAuthorizationNotification];   
}
```

SHKFacebook.m:

```
- (void)session:(FBSession*)session didLogin:(FBUID)uid  {    
    // Try to share again
    if (pendingFacebookAction == SHKFacebookPendingLogin) {
        [self postAuthorizationNotification];   

        self.pendingFacebookAction = SHKFacebookPendingNone;
        [self share];
    }
}
```

**Comment by anthonycmain** [comment]
Fantastic mod, I used the following code to handle the notifications for those that are struggling:
- (void)viewDidLoad {
  [super viewDidLoad];
  
  // Register notification handlers for ShareKit
  [[NSNotificationCenter defaultCenter] addObserver:self 
                                           selector:@selector(shareKitLoginNotification:) 
                                               name:@"SHKLogin" object:nil];

}
- (void)shareKitLoginNotification:(id)object {
  
  if ([[object object] isKindOfClass:[SHKFacebook class]]) {
      // Facebook logged in
      NSLog(@"Facebook logged in %@", object);
  }
  else if ([[object object] isKindOfClass:[SHKTwitter class]]) {
      // Twitter logged in
      NSLog(@"Twitter logged in %@", object);
  }

}

**Comment by miakacham** [comment]
I wonder what's wrong, I followed the steps all the way, but it seems that ShareKitLoginNotification isn't called. :(

**Comment by foreveryh** [comment]
wonderful , i just look for it. Thank you.

**Comment by fwoop** [comment]
i'm sorry for a noob question but, im confused on how to retrieve the twitter username after log in?

**Comment by anthonycmain** [comment]
You will need to listen for a notification (add this code to your viewdidload):

[[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(shareKitLoginNotification:) 
                                                 name:@"SHKTwitterNameRetrieved" object:nil];

And handle it with this method:
- (void)shareKitLoginNotification:(id)object {
  
  [self refreshLabels];
  
  if ([[object object] isKindOfClass:[SHKFacebook class]]) {
      // Facebook logged in
      [self.connectToFacebookButton setTitle:@"Logout   " forState:UIControlStateNormal];
  
  }
  else if ([[object object] isKindOfClass:[SHKTwitter class]]) {
      // Twitter logged in
      [self.connectToTwitterButton setTitle:@"Logout   " forState:UIControlStateNormal];
  
  }  
  else if ([[object name] isEqualToString:@"SHKTwitterNameRetrieved"]) {
          // Twitter logged in
  
  ```
  self.twitterNameLabel.text = [[object userInfo] objectForKey:@"name"];
  ```
  
  }
  else if ([[object name] isEqualToString:@"SHKFacebookNameRetrieved"]) {
      // Facebook logged in
  
  ```
  self.facebookNameLabel.text = [[object userInfo] objectForKey:@"name"];
  ```
  
  }  
  }

**Comment by fwoop** [comment]
@anthonycmain 

Thanks for the reply! but i'm kinda confuse, please take a look at my code.

on my viewDidLoad , i got this :

[[NSNotificationCenter defaultCenter] addObserver:self
selector:@selector(shareKitLoginNotification:) 
name:@"SHKTwitterNameRetrieved" object:nil];

then on the method:
- (void)shareKitLoginNotification:(id)object {
  NSLog(@"%@",object);
  if ([[object object] isKindOfClass:[SHKTwitter class]]) {
  
  ```
  NSLog(@"rawr");
  ```
  
  }
  
  else if ([[object name] isEqualToString:@"SHKTwitterNameRetrieved"]) {
  
  ```
  //self.twitterNameLabel.text = [[object userInfo] objectForKey:@"SHKTwitterNameRetrieved"];
  NSLog(@"%@",[[object userInfo] objectForKey:@"name"]);
  ```
  
  }

}

i use this on my button to authorize:

```
[[[SHKTwitter alloc] init] authorize];
[SHKTwitter release];
```

but for some reason, there is no log on the object.

unless i use SHKLogin.

[[NSNotificationCenter defaultCenter] addObserver:self 
selector:@selector(shareKitLoginNotification:) 
name:@"SHKLogin" object:nil];

and if i do, i get this object on the log

NSConcreteNotification 0x64c8850 {name = SHKLogin; object = <SHKTwitter: 0x6159b50>; userInfo = {
    sharerId = SHKTwitter;
}}

what am i missing? why is there no "name" on the userInfo too

and thanks again for the reply!

**Comment by anthonycmain** [comment]
You must be using a different version of the library to me, looking at the
object you have back you are receiving a userinfo just like I was, try
getting the name field from that dictionary and see if it is stored in there

On 26 July 2012 11:04, fwoop <
reply@reply.github.com

**Comment by fwoop** [comment]
different version of the library of sharekit? im using this http://getsharekit.com/install
sorry, im still really new on iphone / objective C programming. im not sure how to access the dictionary that you are sayng.

**Comment by anthonycmain** [comment]
Yes this post (my version) has updated code to support these features. But also ShareKit itself is now obsolete and is being ran by the community:

https://github.com/ShareKit/ShareKit

But try:

[[object userInfo] objectForKey:@"name"]

**Comment by fwoop** [comment]
darn, im having a lot of problems if i use your version of sharekit. maybe because im still using IOS 4 :(

**Comment by fwoop** [comment]
i know this is too much to ask but can you check my version of sharekit? and add the necessary methods needed to retrieve the twitter username? i'm using IOS 4.

heres my sharekit. http://www.2shared.com/file/H-7xLhg5/ShareKit.html

i'll be forever greatfull @anthonycmain 

also, can you please show on where the areas you have added or edited? i really want to learn :)

**Comment by fwoop** [comment]
btw, if i use [[object userInfo] objectForKey:@"name"] i get a null :(
